tools/clang/lib/Parse/ParseDeclCXX.cpp

```diff
@@ -1580,7 +1580,9 @@ void Parser::ParseClassSpecifier(tok::TokenKind TagTokKind,
 
   if (!Name && !TemplateId && (DS.getTypeSpecType() == DeclSpec::TST_error ||
                                TUK != Sema::TUK_Definition)) {
-    if (DS.getTypeSpecType() != DeclSpec::TST_error) {
+    if (getLangOpts().HLSL && TagTokKind == tok::kw_union) {
+      Diag(Tok, diag::err_anonymous_unions);
```

Comment:
I'm not sure this is the right place for this check. `Sema::BuildAnonymousStructOrUnion` seems more appropriate. You can basically just add `if (getLangOpts.HLSL) { Invalid=true; Diag(...); }` around line 4108, and it should do the right thing.

Reply:
I tried this change, but anonymous unions in HLSL don't hit this function. I understood your motive, though. Ill see what I can do here.